When downloading a virtual good,  check if the order was made by the current user

NMAC427 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | See below.
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Download a virtual good without being logged in / without specifying the order_id or secure_key url parameter. 

Previously anyone with the correct link (correct &key=... URL parameter) could download a virtual good. Specifically:
- If no user was logged in, and either no order ID or no secure key was specified.
- If a user was logged in, but they used a download link from another user.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12023)
<!-- Reviewable:end -->

NMAC427 committed:
Fix invalid URL encoding

Due to the wrong URL encoding the key parameter was lost during the redirect after authentication.

NMAC427 committed:
Check if the order was made by the current user

Previously anyone with the correct link could download a virtual good. Specifically:
- If no user was logged in, and either no order ID or no secure key was specified.
- If a user was logged in, but they used a download link from another user.

NMAC427 committed:
Fix indentation.

NMAC427 commented:
Can someone please help me with resolving the CI linter errors? I can't find any problem... Even running the `PrestaShop/.github/contrib/pre-commit` script doesn't report an error.

matks commented:
Hi @garnele007 and thanks for the contribution 😉 

In order to solve CI lint errors, you can run php-cs-fixer (you need to install composer dev dependencies before): `php ./vendor/bin/php-cs-fixer fix`

See https://devdocs.prestashop.com/1.7/development/coding-standards/ for more details 😉

NMAC427 committed:
Conform to coding standards

NMAC427 commented:
Thanks. Now it's passing the linter. Using `php-cs-fixer` didn't fix the problem. I had to manually remove some trailing whitespace.

Quetzacoalt91 commented:
Hi @garnele007,

By looking at the code you deleted, I wonder if this wasn't expected actually. What happens if the order was made by a guest account?

NMAC427 committed:
Handle guest customers

Fixes a bug where guests weren't able to download virtual products.

NMAC427 committed:
Display error message

If only either the order id or the secure key are set, display an error message.

NMAC427 committed:
Fix indentation

NMAC427 commented:
Hi @Quetzacoalt91,

True. I didn't think about guests. I just made a commit that fixes this. Now my previously committed code only runs if the order was made by a customer with an account.

Quetzacoalt91 commented on controllers/front/GetFileController.php:
```diff
@@ -88,6 +92,15 @@ public function init()
+            // If the order was made by a guest, skip this step
+            $customer = new Customer((int) $order->id_customer);
+            if (!$customer->is_guest) {
+                if ($order->secure_key != $this->context->customer->secure_key) {
```
Could you please combine the 2 if together?

```php
            if (!$customer->is_guest && $order->secure_key !== $this->context->customer->secure_key) {
```

Thanks

NMAC427 committed:
Merge if statements

Quetzacoalt91 commented:
Thank you @garnele007
